src/packet/ipv4.rs

```diff
@@ -39,29 +41,67 @@ pub struct Ipv4 {
     payload: Vec<u8>,
 }
 
-/// Calculates the checksum of an IPv4 packet
-pub fn checksum(packet: &Ipv4Packet) -> u16be {
-    use packet::Packet;
+/// Represents an IPv4 Packet
+///
+/// 0          8          16          24         32
+/// +---------------------------------------------+
+/// |                  source_ip                  |
+/// +----------------------+----------------------+
+/// |                destination_ip               |
+/// +-----------+----------+----------------------+
+/// |   zeros   | protocol | tcp/udp length       |
+/// +-----------+---------------------------------+
+#[packet]
+pub struct Ipv4PseudoHeader {
+    #[construct_with(u8, u8, u8, u8)]
+    source: Ipv4Addr,
+    #[construct_with(u8, u8, u8, u8)]
+    destination: Ipv4Addr,
+    zeros: u8,
+    #[construct_with(u8)]
+    next_level_protocol: IpNextHeaderProtocol,
+    inner_packet_length: u16be,
+    // just to avoid the compiler to complain
+    #[payload]
+    payload: Vec<u8>,
+}
 
-    let len = packet.get_header_length() as usize * 4;
-    let mut sum = 0u32;
-    let mut i = 0;
-    while i < len {
-        let word = (packet.packet()[i] as u32) << 8 | packet.packet()[i + 1] as u32;
-        sum = sum + word;
-        i = i + 2;
-    }
-    while sum >> 16 != 0 {
-        sum = (sum >> 16) + (sum & 0xFFFF);
+impl<'p> PseudoHeader for Ipv4Packet<'p> {
+
+    /// Return a PseudoHeader packet out of the packet header.
+    ///
+    /// The `inner_packet_length` is optional, since in case of UDP, it is taken from the length
+    /// field of the UDP header, whereas for TCP, it is computed from the header length.
+    fn get_pseudo_header(&self, inner_packet_length: Option<u32>) -> Vec<u8> {
```

Comment:
You won't be able to return an `Ipv4PseudoHeader` here, at least not with the current method signature.

The issue is that you're allocating a new vector, which the `get_pseudo_header()` function then owns. Either it needs to return it, so that the caller then owns it, or it needs to be be dropped at the end of the function.

The way around this (and what I'd like you to do - the heap allocation is unfortunate and unnecessary) is something like (untested, but it gives you the general idea):

```
fn populate(&self, buffer: &mut [u8], inner_packet_length: Option<usize>) -> Ipv4PseudoHeader; 
```

Here, the caller allocates the buffer, which the function then populates. Since the caller owns the buffer, its lifetime is managed by the caller, and the return value can outlive the function.